apps/aechannel/src/aesc_session_noise.erl

```diff
@@ -139,35 +145,53 @@ establish({accept, Port, Opts}, St) ->
 establish({connect, Host, Port, Opts}, St) ->
     TcpOpts = tcp_opts(connect, Opts),
     lager:debug("connect: TcpOpts = ~p", [TcpOpts]),
-    {ok, TcpSock} = connect_tcp(Host, Port, TcpOpts),
+    {ok, TcpSock} =
+        connect_tcp(Host, Port, St#st.parent_mon_ref, TcpOpts),
+    lager:debug("Connect TcpSock = ~p", [TcpSock]),
     %% TODO: extract/check something from FinalState?
     EnoiseOpts = enoise_opts(connect, Opts),
     lager:debug("EnoiseOpts (connect) = ~p", [EnoiseOpts]),
     {ok, EConn, _FinalSt} = enoise:connect(TcpSock, EnoiseOpts),
     %% tell_fsm({accept, EConn}, St),
     St#st{econn = EConn}.
 
-connect_tcp(Host, Port, Opts) ->
+connect_tcp(Host, Port, Ref, Opts) ->
     {Timeout, Retries} = get_reconnect_params(),
-    connect_tcp(Retries, Host, Port, Opts, Timeout).
+    connect_tcp(Retries, Host, Port, Ref, Opts, Timeout).
 
-connect_tcp(0, _, _, _, _) ->
+connect_tcp(0, _, _, _, _, _) ->
     erlang:error(connect_timeout);
-connect_tcp(Retries, Host, Port, Opts, Timeout) when Retries > 0 ->
+connect_tcp(Retries, Host, Port, Ref, Opts, Timeout)
+  when Retries > 0 ->
     case gen_tcp:connect(Host, Port, Opts, Timeout) of
         {ok, _TcpSock} = Ok ->
             Ok;
         {error, _} ->
-            timer:sleep(1000),
-            connect_tcp(Retries-1, Host, Port, Opts, Timeout)
+            sleep(Timeout, Ref),
+            connect_tcp(Retries-1, Host, Port, Ref, Opts, Timeout)
     end.
 
+sleep(T, Ref) ->
+    receive
+        {'$gen_call', From, close} = Msg ->
+            lager:debug("Got ~p while sleeping", [Msg]),
+            gen:reply(From, ok),
+            exit(normal);
+        {'DOWN', Ref, _, _, _} = Msg ->
+            lager:debug("Got ~p while sleeping", [Msg]),
+            exit(shutdown)
+    after T ->
+            ok
+    end.
+
+
 get_reconnect_params() ->
     %% {ConnectTimeout, Retries}
-    {10000, 30}.
+    {3000, 40}.  %% total 2 minutes
```

Comment:
Timeout and number of retries may be made user configurable though seems ok for now.